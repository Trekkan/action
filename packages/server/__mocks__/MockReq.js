export default class MockRes {
  constructor (options = {}) {
    const {headers, body} = options
    this.headers = headers || {}
    this.body = body
    // we mock everything that uses rawBody for performance
    this.rawBody = body
  }
  get (header) {
    return this.headers[header]
  }
}
