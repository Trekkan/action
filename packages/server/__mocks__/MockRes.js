/* eslint-env jest */

export default class MockRes {
  constructor () {
    this.sendStatus = jest.fn()
  }
}
