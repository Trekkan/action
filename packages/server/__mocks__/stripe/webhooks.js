export default {
  constructEvent (body) {
    return body;
  }
};
