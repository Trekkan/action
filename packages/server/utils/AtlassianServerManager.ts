import fetch from 'node-fetch'
import makeAppLink from './makeAppLink'
import AtlassianManager from 'parabol-client/utils/AtlassianManager'

interface AuthQueryParams {
  grant_type: 'authorization_code'
  code: string
  redirect_uri: string
}

interface RefreshQueryParams {
  grant_type: 'refresh_token'
  refresh_token: string
}

interface OAuth2Response {
  access_token: string
  refresh_token: string
  error: any
  scope: string
}

class AtlassianServerManager extends AtlassianManager {
  fetch = fetch
  static async init(code: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'authorization_code',
      code,
      redirect_uri: makeAppLink('auth/atlassian')
    })
  }

  static async refresh(refreshToken: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'refresh_token',
      refresh_token: refreshToken
    })
  }

  static async fetchToken(partialQueryParams: AuthQueryParams | RefreshQueryParams) {
    const queryParams = {
      ...partialQueryParams,
      client_id: process.env.ATLASSIAN_CLIENT_ID,
      client_secret: process.env.ATLASSIAN_CLIENT_SECRET
    }

    const uri = `https://auth.atlassian.com/oauth/token`

    const tokenRes = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(queryParams)
    })

    const tokenJson = (await tokenRes.json()) as OAuth2Response

    const {access_token: accessToken, refresh_token: refreshToken, error, scope} = tokenJson
    if (error) {
      throw new Error(`Atlassian: ${error}`)
    }
    const providedScope = scope.split(' ')
    const matchingScope =
      new Set([...AtlassianServerManager.SCOPE.split(' '), ...providedScope]).size ===
      providedScope.length
    if (!matchingScope) {
      throw new Error(`bad scope: ${scope}`)
    }
    return new AtlassianServerManager(accessToken, refreshToken)
  }

  refreshToken?: string

  constructor(accessToken: string, refreshToken?: string) {
    super(accessToken)
    this.refreshToken = refreshToken
  }
}

export default AtlassianServerManager
