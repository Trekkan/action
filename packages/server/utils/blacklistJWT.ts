import getBlacklistJWTKey from './getBlacklistJWTKey'
import {Threshold, SubscriptionChannel} from 'parabol-client/types/constEnums'
import {toEpochSeconds} from './epochTime'
import publish from './publish'
import getRedis from './getRedis'

const blacklistJWT = async (userId: string, iat: number, mutatorId?: string) => {
  const key = getBlacklistJWTKey(userId)
  const redis = getRedis()
  const expiresIn = iat - toEpochSeconds(new Date()) + Math.floor(Threshold.JWT_LIFESPAN / 1000)
  await redis.set(key, iat, 'EX', expiresIn)
  publish(SubscriptionChannel.NOTIFICATION, userId, 'InvalidateSessionsPayload', {}, {mutatorId})
}

export default blacklistJWT
