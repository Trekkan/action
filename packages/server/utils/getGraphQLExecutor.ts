import {ExecutionResult} from 'graphql'
import {ServerChannel} from 'parabol-client/types/constEnums'
import PubSubPromise from './PubSubPromise'

export interface PubSubPromiseResponse extends ExecutionResult {
  jobId: string
}

let pubsub: PubSubPromise<PubSubPromiseResponse>
const getGraphQLExecutor = () => {
  if (!pubsub) {
    pubsub = new PubSubPromise(
      ServerChannel.GQL_EXECUTOR_REQUEST,
      ServerChannel.GQL_EXECUTOR_RESPONSE
    )
  }
  return pubsub
}
export default getGraphQLExecutor
