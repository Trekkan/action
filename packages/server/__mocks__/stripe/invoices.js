import shortid from 'shortid';
import {deletedReturnVal, getDoc, updateFromOptions} from './utils';
import creditCardByToken from 'server/__tests__/utils/creditCardByToken';
import {toEpochSeconds} from 'server/utils/epochTime';
import {makeSubscriptionPlan} from './subscriptions';

export const makeInvoiceLine = (overrides) => {
  const now = new Date();
  const nowInSeconds = toEpochSeconds(now);
  const {id, metadata, periodStart, subscription} = overrides;
  return {
    "id": id,
    "object": "line_item",
    "amount": 0,
    "currency": "usd",
    "description": null,
    "discountable": true,
    "livemode": true,
    "metadata": metadata,
    "period": {
      "start": periodStart || 1488909033,
      "end": 1491501033
    },
    "plan": makeSubscriptionPlan(nowInSeconds),
    "proration": false,
    "quantity": 1,
    "subscription": subscription,
    "subscription_item": `si_${shortid.generate()}`,
    "type": "subscription"
  }
};

const makeNextMonthChargesLine = (subscription) => makeInvoiceLine({
  id: subscription.id,
  metadata: {
    orgId: subscription.metadata.orgId
  },
});

const makeInvoiceLines = (id, subscription, lineTypes) => {
  const totalCount = lineTypes ? Object.keys(lineTypes).reduce((count, type) => count + lineTypes[type], 0) : undefined;
  return {
    "data": [
      makeNextMonthChargesLine(subscription),
    ],
    "total_count": totalCount,
    "object": "list",
    "url": `/v1/invoices/${id}/lines`
  }
};

export const createNewInvoice = (overrides, reject) => {
  const {id, customer, subscription} = overrides;
  return {
    "id": id,
    "object": "invoice",
    "amount_due": 0,
    "application_fee": null,
    "attempt_count": 0,
    "attempted": true,
    "charge": null,
    "closed": true,
    "currency": "usd",
    "customer": customer,
    "date": toEpochSeconds(new Date()),
    "description": null,
    "discount": null,
    "ending_balance": 0,
    "forgiven": false,
    "lines": makeInvoiceLines(id, subscription),
    "livemode": false,
    "metadata": {},
    "next_payment_attempt": null,
    "paid": true,
    "period_end": 1488909033,
    "period_start": 1488909033,
    "receipt_number": null,
    "starting_balance": 0,
    "statement_descriptor": null,
    "subscription": subscription.id,
    "subtotal": 0,
    "tax": null,
    "tax_percent": null,
    "total": 0,
    "webhooks_delivered_at": 1488909040
  }
};

export default (stripe) => ({
   create: jest.fn((options) => new Promise((resolve, reject) => {
     const {id} = options;
     const invoiceDoc = stripe.__db.invoices[id] = createNewInvoice(options, reject);
     resolve(invoiceDoc);
   })),
  retrieve: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    resolve(doc);
  })),
  retrieveLines: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    resolve(doc.lines);
  })),
  //TODO
  retrieveUpcoming: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    resolve(doc);
  })),
  //
  update: jest.fn((id, options) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    updateFromOptions(doc, stripe.invoices.__updateHandlers, options, reject);
    resolve(doc)
  })),
  del: jest.fn((id) => new Promise((resolve, reject) => {
    getDoc(stripe.__db.invoices[id], reject);
    delete stripe.__db.invoices[id];
    resolve(deletedReturnVal(id));
  })),
  __trimFields: [
    'id',
    'customer',
    'metadata.orgId',
    'subscription',
    'lines.url',
    'lines.data.id',
    'lines.data.metadata.orgId',
    'lines.data.subscription_item'
  ],
  __triggers: ['update', 'del'],
  __uniqueKeyField: 'customer',
  __updateHandlers: {}
});
