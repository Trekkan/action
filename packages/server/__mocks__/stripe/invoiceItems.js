import {getDoc, updateFromOptions} from './utils';
import {makeInvoiceLine} from './invoices';

export default (stripe) => ({
  create: jest.fn((options) => new Promise((resolve) => {
    const {id} = options;
    const invoiceDoc = stripe.__db.invoiceItems[id] = makeInvoiceLine(options);
    resolve(invoiceDoc);
  })),
  retrieve: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoiceItems[id], reject);
    resolve(doc);
  })),
  update: jest.fn((id, options) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoiceItems[id], reject);
    updateFromOptions(doc, stripe.invoiceItems.__updateHandlers, options, reject);
    resolve(doc)
  })),
  __trimFields: [
    'id',
    'customer',
    'metadata.userId',
    'subscription',
    'subscription_item'
  ],
  __triggers: ['update'],
  __uniqueKeyField: 'subscription',
  __updateHandlers: {}
});