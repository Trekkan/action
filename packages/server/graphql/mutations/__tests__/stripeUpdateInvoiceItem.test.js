/* eslint-env jest */
import DynamicSerializer from 'dynamic-serializer'
import MockReq from '../../../__mocks__/MockReq'
import MockRes from '../../../__mocks__/MockRes'
import MockDB from '../../../__tests__/setup/MockDB'
import fetchAndSerialize from '../../../__tests__/utils/fetchAndSerialize'
import stripe from '../../../billing/stripe'
import stripeWebhookHandler from '../../../billing/stripeWebhookHandler'
import DataLoaderWarehouse from 'dataloader-warehouse'
import getRethink from '../../../database/rethinkDriver'
import invoiceItemCreatedEvent from './mockStripeEvents/invoiceItemCreatedEvent'
import shortid from 'shortid'
import { PRO } from 'parabol-client/utils/constants'

console.error = jest.fn()

describe('stripeUpdateInvoiceItem', () => {
  test('handles invoiceitem.created', async () => {
    // SETUP
    const hookId = shortid.generate()
    const invoiceItemId = shortid.generate()
    const res = new MockRes()
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockDB = new MockDB()
    const { organization, invoiceItemHook } = await mockDB
      .init({ plan: PRO })
      .newInvoiceItemHook({ id: hookId })

    const org = organization[0]
    const hook = invoiceItemHook[0]
    stripe.__setMockData(org)
    await stripe.invoiceItems.create({
      subscription: org.stripeSubscriptionId,
      id: invoiceItemId,
      metadata: { orgId: org.id },
      periodStart: hook.prorationDate
    })
    const req = new MockReq({
      body: invoiceItemCreatedEvent(invoiceItemId, org.stripeId, org.stripeSubscriptionId)
    })

    // TEST
    const sharedDataLoader = new DataLoaderWarehouse({
      ttl: 1000,
      onShare: '_share'
    })
    await stripeWebhookHandler(sharedDataLoader)(req, res)

    // VERIFY
    expect(res.sendStatus).toBeCalledWith(200)
    const db = await fetchAndSerialize(
      {
        invoiceItemHook: r.table('InvoiceItemHook').get(hookId)
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(stripe.__snapshot(org.stripeSubscriptionId, dynamicSerializer)).toMatchSnapshot()
  })
})
