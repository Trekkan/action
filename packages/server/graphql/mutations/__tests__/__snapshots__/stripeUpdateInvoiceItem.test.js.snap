// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripeUpdateInvoiceItem handles invoiceitem.created 1`] = `
Object {
  "invoiceItemHook": Object {
    "id": 0,
    "prorationDate": 1483254000,
    "stripeSubscriptionId": 1,
    "type": "addUser",
    "userId": 2,
  },
}
`;

exports[`stripeUpdateInvoiceItem handles invoiceitem.created 2`] = `
Object {
  "invoiceItems": Array [
    Object {
      "amount": 0,
      "currency": "usd",
      "description": null,
      "discountable": true,
      "id": 3,
      "livemode": true,
      "metadata": Object {
        "type": "addUser",
        "userId": 2,
      },
      "object": "line_item",
      "period": Object {
        "end": 1491501033,
        "start": 1483254000,
      },
      "plan": Object {
        "amount": 500,
        "created": 1483257600,
        "currency": "usd",
        "id": "parabol-pro-monthly",
        "interval": "month",
        "interval_count": 1,
        "livemode": false,
        "metadata": Object {},
        "name": "parabol-pro-monthly",
        "object": "plan",
        "statement_descriptor": null,
        "trial_period_days": null,
      },
      "proration": false,
      "quantity": 1,
      "subscription": 1,
      "subscription_item": 4,
      "type": "subscription",
    },
  ],
}
`;
